pkgs/dart_mcp_server/lib/src/utils/cli_utils.dart

```diff
@@ -91,103 +91,150 @@ Future<CallToolResult> runCommandInRoots(
-    final rootUriString = rootConfig[ParameterNames.root] as String?;
-    if (rootUriString == null) {
-      // This shouldn't happen based on the schema, but handle defensively.
-      return CallToolResult(
-        content: [
-          TextContent(text: 'Invalid root configuration: missing `root` key.'),
-        ],
-        isError: true,
-      );
-    }
+    final result = await runCommandInRoot(
+      request,
+      rootConfig: rootConfig,
+      commandForRoot: commandForRoot,
+      arguments: arguments,
+      commandDescription: commandDescription,
+      fileSystem: fileSystem,
+      processManager: processManager,
+      knownRoots: knownRoots,
+      defaultPaths: defaultPaths,
+    );
+    if (result.isError == true) return result;
+    outputs.addAll(result.content);
+  }
+  return CallToolResult(content: outputs);
+}
 
-    final root = _findRoot(rootUriString, knownRoots);
-    if (root == null) {
-      return CallToolResult(
-        content: [
-          TextContent(
-            text:
-                'Invalid root $rootUriString, must be under one of the '
-                'registered project roots:\n\n${knownRoots.join('\n')}',
-          ),
-        ],
-        isError: true,
-      );
-    }
+/// Runs [commandForRoot] in a single project root specified in the
+/// [request], with [arguments].
+///
+/// If [rootConfig] is passed, this will be used to read the root configuration,
+/// otherwise it is read directly off of `request.arguments`.
+///
+/// These [commandForRoot] plus [arguments] are passed directly to
+/// [ProcessManager.run].
+///
+/// The [commandDescription] is used in the output to describe the command
+/// being run. For example, if the command is `['dart', 'fix', '--apply']`, the
+/// command description might be `dart fix`.
+///
+/// [defaultPaths] may be specified if one or more path arguments are required
+/// for the command (e.g. `dart format <default paths>`). The paths can be
+/// absolute or relative paths that point to the directories on which the
+/// command should be run. For example, the `dart format` command may pass a
+/// default path of '.', which indicates that every Dart file in the working
+/// directory should be formatted. The value of `defaultPaths` will only be used
+/// if the [request]'s root configuration does not contain a set value for a
+/// root's 'paths'.
+Future<CallToolResult> runCommandInRoot(
+  CallToolRequest request, {
+  Map<String, Object?>? rootConfig,
+  FutureOr<String> Function(Root, FileSystem) commandForRoot =
+      defaultCommandForRoot,
+  List<String> arguments = const [],
+  required String commandDescription,
+  required FileSystem fileSystem,
+  required ProcessManager processManager,
+  required List<Root> knownRoots,
+  List<String> defaultPaths = const <String>[],
+}) async {
+  rootConfig ??= request.arguments;
+  final rootUriString = rootConfig?[ParameterNames.root] as String?;
+  if (rootUriString == null) {
```

Comment:
nit: we could check `rootConfig?.containsKey(ParameterNames.root) ?? false` instead